src/csharp/Grpc.Core/ContextPropagationToken.cs

```diff
@@ -32,124 +26,10 @@ namespace Grpc.Core
-        readonly DateTime deadline;
-        readonly CancellationToken cancellationToken;
-        readonly ContextPropagationOptions options;
-
-        internal ContextPropagationToken(CallSafeHandle parentCall, DateTime deadline, CancellationToken cancellationToken, ContextPropagationOptions options)
-        {
-            this.parentCall = GrpcPreconditions.CheckNotNull(parentCall);
-            this.deadline = deadline;
-            this.cancellationToken = cancellationToken;
-            this.options = options ?? ContextPropagationOptions.Default;
-        }
-
-        /// <summary>
-        /// Gets the native handle of the parent call.
-        /// </summary>
-        internal CallSafeHandle ParentCall
-        {
-            get
-            {
-                return this.parentCall;
-            }
-        }
-
-        /// <summary>
-        /// Gets the parent call's deadline.
-        /// </summary>
-        internal DateTime ParentDeadline
-        {
-            get
-            {
-                return this.deadline;
-            }
-        }
-
-        /// <summary>
-        /// Gets the parent call's cancellation token.
-        /// </summary>
-        internal CancellationToken ParentCancellationToken
-        {
-            get
-            {
-                return this.cancellationToken;
-            }
-        }
-
-        /// <summary>
-        /// Get the context propagation options.
-        /// </summary>
-        internal ContextPropagationOptions Options
-        {
-            get
-            {
-                return this.options;
-            }
-        }
-    }
-
-    /// <summary>
-    /// Options for <see cref="ContextPropagationToken"/>.
-    /// </summary>
-    public class ContextPropagationOptions
+    public abstract class ContextPropagationToken
     {
-        /// <summary>
-        /// The context propagation options that will be used by default.
-        /// </summary>
-        public static readonly ContextPropagationOptions Default = new ContextPropagationOptions();
-
-        bool propagateDeadline;
-        bool propagateCancellation;
-
-        /// <summary>
-        /// Creates new context propagation options.
-        /// </summary>
-        /// <param name="propagateDeadline">If set to <c>true</c> parent call's deadline will be propagated to the child call.</param>
-        /// <param name="propagateCancellation">If set to <c>true</c> parent call's cancellation token will be propagated to the child call.</param>
-        public ContextPropagationOptions(bool propagateDeadline = true, bool propagateCancellation = true)
-        {
-            this.propagateDeadline = propagateDeadline;
-            this.propagateCancellation = propagateCancellation;
-        }
-            
-        /// <summary><c>true</c> if parent call's deadline should be propagated to the child call.</summary>
-        public bool IsPropagateDeadline
+        internal ContextPropagationToken()
```

Comment:
Does it make sense for `Grpc.AspNetCore` (or anyone else) to create their own token? internal will prevent other assemblies from extending. If every other implementation can return null from `CreatePropagationTokenCore` then this is fine.

Reply:
Grpc.AspNetCore.Server should have some notion of propagating the context (as this is generally useful feature, important for things like tracing), but I don't think we can easily make it work when Grpc.AspNetCore.Server and Grpc.Core are used in the same process (= before the pure c# client gets implemented).

Ad "internal" visibility: Grpc.AspNetCore.Server can get access to Grpc.Core.Api internals via [InternalsVisibleTo] attribute. We could also make the constructor protected, but I'm trying not to expose publicly more than necessary. 

Until Grpc.AspNetCore.Server provides some useful functionality in terms of propagating context, I think it's fine for Grpc.AspNetCore.Server's ServerCallContext.CreateContextPropagationToken() to just return null (or even better to throw `NotImplementedException`).

Reply:
Filed tracking issue: https://github.com/grpc/grpc-dotnet/issues/40.

Reply:
Note that InternalVisibleTo is not a great solution due to the coupling involved. For example, when the Microsoft.AspNetCore.Server package becomes strong named, we'll need to remember to add the public key to the attribute. I am in favour of making the constructor protected, but we could live without this feature for now.